src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -239,26 +239,27 @@ private class TaskLauncherActor(
   }
 
   private[this] def receiveInstanceUpdate: Receive = {
-    case update: InstanceDeleted =>
-      logger.info(s"receiveInstanceUpdate: ${update.id} was deleted (${update.condition})")
-      // A) If the app has constraints, we need to reconsider offers that
-      // we already rejected. E.g. when a host:unique constraint prevented
-      // us to launch tasks on a particular node before, we need to reconsider offers
-      // of that node after a task on that node has died.
-      //
-      // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
-      if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
-        maybeOfferReviver.foreach(_.reviveOffers())
+    case update: InstanceUpdated =>
+      if (update.condition.isTerminal & update.instance.isScheduled) {
+        logger.info(s"receiveInstanceUpdate: ${update.id} is terminal (${update.condition}) and scheduled.")
+        // A) If the app has constraints, we need to reconsider offers that
+        // we already rejected. E.g. when a host:unique constraint prevented
+        // us to launch tasks on a particular node before, we need to reconsider offers
+        // of that node after a task on that node has died.
+        //
+        // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
+        if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
+          maybeOfferReviver.foreach(_.reviveOffers())
+        }
       }
       syncInstance(update.instance.instanceId)
       OfferMatcherRegistration.manageOfferMatcherStatus()
       sender() ! Done
 
-    case change: InstanceChange =>
-      syncInstance(change.instance.instanceId)
-      OfferMatcherRegistration.manageOfferMatcherStatus()
-      sender() ! Done
-
+    case update: InstanceDeleted =>
+      // if an instance was deleted, it's not needed anymore and we only have to remove it from the internal state
+      logger.info(s"${update.instance.instanceId} was deleted. Will remove from internal state.")
+      removeInstanceFromInternalState(update.instance.instanceId)
```

Comment:
Whenever an instance is updated (the `case update: InstanceUpdated` above), we do
```
syncInstance(update.instance.instanceId)
OfferMatcherRegistration.manageOfferMatcherStatus()
```
So when the goal is adjusted, we do this. But I think in the case when an instance is terminal (e.g. a task failed before and we don't get matching offers) and the goal is adjusted to Decommissioned, it will immediately be deleted, so you're right, we should probably do this here as well.

Reply:
> shouldn't we call offer matcher registration, just in case?

We are calling `OfferMatcherRegistration.manageOfferMatcherStatus()` just below, right?

Reply:
just out of curiosity, why did we remove the `syncInstance`? It was doing just what we needed. I'm surprised that we touch the `TaskLauncherActor` at all.